hack/imagebuilder/main.go

```diff
@@ -0,0 +1,296 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"bufio"
+	"flag"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"sync"
+
+	"context"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	defaultArch = "linux/amd64"
+	allArch     = "all"
+
+	gatherStaicScriptName = "gather-static.sh"
+
+	// Relative to root of the repo
+	defaultProwImageListFile = "prow/.prow-images.yaml"
+
+	defaultWorkersCount = 10
+
+	// noOpKoDocerRepo is used when images are not pushed
+	noOpKoDocerRepo = "do.not/matter/at/all"
+)
+
+var (
+	rootDir     string
+	otherArches = []string{
+		"arm64",
+		"s390x",
+		"ppc64le",
+	}
+	defaultTags = []string{
+		"latest",
+		"latest-root",
+	}
+)
+
+func init() {
+	out, err := runCmd("git", "rev-parse", "--show-toplevel")
+	if err != nil {
+		logrus.WithError(err).Error("Failed getting git root dir")
+		os.Exit(1)
+	}
+	rootDir = out
+}
+
+func init() {
+	_, err := runCmdInDir(path.Join(rootDir, "hack/tools"), "go", "build", "-o", path.Join(rootDir, "_bin/ko"), "github.com/google/ko")
+	if err != nil {
+		logrus.WithError(err).Error("Failed ensure ko")
+		os.Exit(1)
+	}
+}
+
+type options struct {
+	koDockerRepo      string
+	prowImageListFile string
+	workers           int
+	push              bool
+}
+
+func runCmdInDir(dir, cmd string, args ...string) (string, error) {
+	command := exec.Command(cmd, args...)
+	if dir != "" {
+		command.Dir = dir
+	}
+	stdOut, err := command.StdoutPipe()
+	if err != nil {
+		return "", err
+	}
+	stdErr, err := command.StderrPipe()
+	if err != nil {
+		return "", err
+	}
+	if err := command.Start(); err != nil {
+		return "", err
+	}
+	scanner := bufio.NewScanner(stdOut)
+	var allOut string
+	for scanner.Scan() {
+		out := scanner.Text()
+		allOut = allOut + out
+		logrus.WithField("cmd", command.Args).Info(out)
+	}
+	allErr, _ := io.ReadAll(stdErr)
+	if len(allErr) > 0 {
+		logrus.WithField("cmd", command.Args).Error(string(allErr))
+	}
+	err = command.Wait()
+	return strings.TrimSpace(allOut), err
+}
+
+func runCmd(cmd string, args ...string) (string, error) {
+	return runCmdInDir(rootDir, cmd, args...)
+}
+
+type imageDef struct {
+	Dir  string `json:"dir"`
+	Arch string `json:"arch"`
+}
+
+type imageDefs struct {
+	Defs []imageDef `json:"images"`
+}
+
+func loadImageDefs(p string) ([]imageDef, error) {
+	b, err := ioutil.ReadFile(p)
+	if err != nil {
+		return nil, err
+	}
+	var res imageDefs
+	if err := yaml.Unmarshal(b, &res); err != nil {
+		return nil, err
+	}
+	return res.Defs, nil
+}
+
+func allBaseTags() ([]string, error) {
+	gitTag, err := gitTag()
+	if err != nil {
+		return nil, err
+	}
+	return append(defaultTags, gitTag), nil
+}
+
+func allTags(arch string) ([]string, error) {
+	baseTags, err := allBaseTags()
+	if err != nil {
+		return nil, err
+	}
+	if arch != allArch {
+		return baseTags, nil
+	}
+	var allTags = baseTags[:]
```

Comment:
This is the same as `allTags := baseTags` right?

Reply:
yeah, this was deep clone in python :)